OpenRA.Mods.Common/Traits/BotModules/McvExpansionManagerBotModule.cs

```diff
@@ -0,0 +1,767 @@
+						&& Info.ConstructionYardTypes.Contains(a.Info.Name)).Select(a => (a.Location, a.Owner != player)).ToArray();
+
+					var crc_refinarylocs = world.ActorsHavingTrait<Refinery>().Where(a => a.Owner.IsAlliedWith(player) && Info.RefineryTypes.Contains(a.Info.Name))
+						.Select(a => (a.Location, a.Owner != player))
+						.ToArray();
+
+					var crc_rescreators = world.ActorsHavingTrait<SeedsResource>().Where(a => Info.ResourceCreatorTypes.Contains(a.Info.Name));
+
+					CPos? crc_suitablelocation = null;
+					var crc_best = int.MinValue;
+
+					foreach (var rescreator in crc_rescreators)
+					{
+						var attraction = indiceStrideSquare >> 3;
+
+						attraction -= (rescreator.Location - mcv_loc).LengthSquared / pathDistanceSquareFactor;
+
+						if (lastFailedExpandSpot == rescreator.Location)
+							continue;
+
+						if (world.FindActorsInCircle(rescreator.CenterPosition, WDist.FromCells(Info.CRCmodeEnemyScanRadius)).Any(a => !a.Disposed
+							&& (player.RelationshipWith(a.Owner) == PlayerRelationship.Enemy)
+							&& a.Info.HasTraitInfo<BuildingInfo>()
+							&& a.Info.HasTraitInfo<SellableInfo>()))
+							attraction -= indiceStrideSquare << 2;
+
+						foreach (var (location, isAlly) in crc_refinarylocs)
+						{
+							var sdistance = (rescreator.Location - location).LengthSquared;
+							if (sdistance <= Info.CRCmodeRefineryUnfavorRange * Info.CRCmodeRefineryUnfavorRange)
+							{
+								if (isAlly)
+									attraction -= indiceStrideSquare;
+								else
+									attraction -= indiceStrideSquare << 1;
+							}
+						}
+
+						foreach (var (location, isAlly) in crc_conyardlocs)
+						{
+							var sdistance = (rescreator.Location - location).LengthSquared;
+							if (sdistance <= Info.CRCmodeConyardUnfavorRange * Info.CRCmodeConyardUnfavorRange)
+							{
+								if (isAlly)
+									attraction -= indiceStrideSquare;
+								else
+									attraction -= indiceStrideSquare << 1;
+							}
+						}
+
+						if (!allowfallback)
+						{
+							var sdistance = (rescreator.Location - mcv_loc).LengthSquared;
+							if (sdistance <= indiceStrideSquare)
+								attraction -= indiceStrideSquare << 1;
+						}
+
+						if (attraction > crc_best)
+						{
+							crc_best = attraction;
+							crc_suitablelocation = rescreator.Location;
+						}
+					}
+
+					return (crc_suitablelocation, crc_best);
+
+				default:
+					return (null, int.MinValue);
+			}
+		}
+
+		void IBotTick.BotTick(IBot bot)
+		{
+			attackrespondcooldown--;
+
+			if (firstTick)
+			{
+				// check which mode we should use in map
+				SwitchExpansionMode(Info.InitialExpansionMode);
+				if (mcvExpansionMode != BotMcvExpansionMode.CheckBase && !world.ActorsHavingTrait<SeedsResource>().Any())
+				{
+					SwitchExpansionMode(BotMcvExpansionMode.CheckResource);
+					if (!world.Map.AllCells.Any(a => resourceTypeIndices.Get(world.Map.GetTerrainIndex(a))))
+						SwitchExpansionMode(BotMcvExpansionMode.CheckBase);
+				}
```

Comment:
```suggestion
				if (Info.InitialExpansionMode == BotMcvExpansionMode.CheckResourceCreator
						&& world.ActorsHavingTrait<SeedsResource>().Any())
					SwitchExpansionMode(BotMcvExpansionMode.CheckResourceCreator);
				else if (Info.InitialExpansionMode != BotMcvExpansionMode.CheckBase
						&& world.Map.AllCells.Any(a => resourceTypeIndices.Get(world.Map.GetTerrainIndex(a))))
					SwitchExpansionMode(BotMcvExpansionMode.CheckResource);
				else
					SwitchExpansionMode(BotMcvExpansionMode.CheckBase);
```

This is more foolproof

Reply:
fixed